app/src/main/java/supermarket/model/Offer.kt

```diff
@@ -1,3 +1,118 @@
 package supermarket.model
 
-class Offer(internal var offerType: SpecialOfferType, internal val product: Product, internal var argument: Double)
+sealed class Offer {
+    abstract fun getDiscount(
+        shoppingCart: ShoppingCart,
+        catalog: SupermarketCatalog
+    ): Discount?
+}
+
+abstract class SingleProductOffer(val product: Product): Offer()
+
+class PercentageOffer(product: Product, val percentOff: Double) : SingleProductOffer(product) {
+    override fun getDiscount(
+        shoppingCart: ShoppingCart,
+        catalog: SupermarketCatalog
+    ): Discount? {
+        val quantityOf = shoppingCart.quantityOf(product)?.toInt() ?: return null
+
+        val unitPrice = catalog.getUnitPrice(product)
+        return Discount(product, "$percentOff% off", quantityOf * unitPrice * percentOff / 100.0)
+    }
+}
+
+class ThreeForTwoOffer(product: Product) : SingleProductOffer(product) {
+    override fun getDiscount(
+        shoppingCart: ShoppingCart,
+        catalog: SupermarketCatalog
+    ): Discount? {
+        TODO("Not yet implemented")
+    }
+}
+
+class QuantityForAmountOffer(product: Product, private val discountQuantity: Int, private val price: Double) :
+    SingleProductOffer(product) {
+    override fun getDiscount(
+        shoppingCart: ShoppingCart,
+        catalog: SupermarketCatalog
+    ): Discount? {
+        val productQuantity = shoppingCart.quantityOf(product)
+        val quantityInt = productQuantity?.toInt()?.let {
```

Comment:
Given what we have I think using an int makes sense. the existing offer types before the refactor the special offer types specifically offered were `ThreeForTwo, TenPercentDiscount, TwoForAmount, FiveForAmount`

It assumed whole numbers for the quantity of amount types. While we could make it so partial quantities could work for this kind of discount, with the existing assumptions I think it's safe to assume whole amounts for this kind of offer. It would also coincide with real life examples I see at stores, which usually are in terms of whole amount quantities.

Reply:
Totally agree that "2.5kg of rice for $5" can make sense but I don't think that's what the intent of this one was originally, as per Cee's comment.